### PR TITLE
feat(network): start network impl and rpc handling

### PR DESCRIPTION
**Pull request by dancoombs**
#248 

## Proposed Changes

  -  Start the `Network` implementation with the RPC behavior (sorry about the behaviour spelling, thats what libp2p uses to I was consistent).
  - Implement the RPC behavior with its custom connection handler, and inbound and outbound protocols
  - Start an integration test suite

### CODE REVIEW

**Comment by dancoombs** [comment]
Next up:

* Gossipsub implementation
* Discv5 discovery behavior
* Peer management
  * Metadata, status, ping management
  * Peer reputation and scoring
  * Peer rate limiting

File: crates/network/src/enr.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            Self::Ed25519(pk) => {
+                let pk_bytes = pk.to_bytes();
+                let libp2p_pk: PublicKey = ed25519::PublicKey::try_from_bytes(&pk_bytes)
+                    .expect("valid public key")
+                    .into();
+                PeerId::from_public_key(&libp2p_pk)
+            }
+        }
+    }
+}
```
Comment (dancoombs) [inline crates/network/src/enr.rs:98]: TODO: tests

File: crates/network/src/network.rs [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+        let local_peer_id = PeerId::from(local_key.public());
+
+        let (enr, _enr_key) = enr::build_enr(&config);
+
+        let mplex_config = libp2p_mplex::MplexConfig::new();
+        let yamux_config = libp2p::yamux::Config::default();
+
+        let transport = tcp::tokio::Transport::default()
+            .upgrade(upgrade::Version::V1Lazy)
+            .authenticate(noise::Config::new(&local_key).unwrap())
```
Comment (dancoombs) [inline crates/network/src/network.rs:186]: TODO: remove these unwraps

File: crates/network/src/network.rs [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+            secp256k1::SecretKey::try_from_bytes(&mut hex::decode(&config.private_key).unwrap())
+                .unwrap(),
+        )
+        .into();
+        let local_peer_id = PeerId::from(local_key.public());
+
+        let (enr, _enr_key) = enr::build_enr(&config);
+
+        let mplex_config = libp2p_mplex::MplexConfig::new();
+        let yamux_config = libp2p::yamux::Config::default();
```
Comment (dancoombs) [inline crates/network/src/network.rs:182]: TODO: pass these in through config

File: bin/tools/src/bin/network.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+// If not, see https://www.gnu.org/licenses/.
+
+use std::{net::SocketAddr, time::Duration};
+
+use clap::Parser;
+use ethers::types::H256;
+use rundler_network::{Config as NetworkConfig, ConnectionConfig, Network};
+use tokio::sync::mpsc;
+
+const PRIVATE_KEY: &str = "b0ddfec7d365b4599ff8367e960f8c4890364f99e2151beac352338cc0cfe1bc";
```
Comment (alex-miao) [inline bin/tools/src/bin/network.rs:21]: move this to a .env

**Pull request closed without merging**